include/pangolin/utils/file_utils.h

```diff
@@ -125,6 +126,78 @@ inline std::string ToLowerCopy( const std::string& str )
     return out;
 }
 
+/**
+ * A subclass of ifstream that (optionally) supports the GNU stdio_filebuf
+ * extension to construct itself from a file descriptor.
+ */
+class ifstream : public std::ifstream
+{
+public:
+    explicit ifstream() :
+        _filebuf(nullptr)
+    {
+    }
+
+    explicit ifstream(const char *filename, std::ios_base::openmode mode = std::ios_base::in) :
+        std::ifstream(filename, mode),
+        _filebuf(nullptr)
+    {
+    }
+
+    explicit ifstream(const std::string& filename, std::ios_base::openmode mode = std::ios_base::in) :
+        std::ifstream(filename, mode),
+        _filebuf(nullptr)
+    {
+    }
+
+    explicit ifstream(int fd, std::ios_base::openmode mode = std::ios_base::in) :
+        _filebuf(nullptr)
+    {
+        open(fd, mode);
+    }
+
+    ~ifstream()
+    {
+        close();
+    }
+
+    void open(const char *filename, std::ios_base::openmode mode = std::ios_base::in)
+    {
+        std::ifstream::open(filename, mode);
+    }
+
+    void open(const std::string& filename, std::ios_base::openmode mode = std::ios_base::in)
+    {
+        std::ifstream::open(filename, mode);
+    }
+
+    void open(int fd, std::ios_base::openmode mode = std::ios_base::in)
+    {
+        if (_filebuf != nullptr) {
+            setstate(std::ifstream::failbit);
+        } else {
+            _filebuf = new __gnu_cxx::stdio_filebuf<std::ifstream::char_type>(fd,
+                mode);
+            static_cast<std::basic_ios<std::ifstream::char_type> *>(this)->rdbuf(_filebuf);
+        }
+    }
+
+    void close()
+    {
+        std::ifstream::close();
+
+        if (_filebuf != nullptr) {
+            if (_filebuf->is_open()) {
+                _filebuf->close();
+            }
+            delete _filebuf;
+            _filebuf = nullptr;
+        }
+    }
+
+private:
+    __gnu_cxx::stdio_filebuf<std::ifstream::char_type> *_filebuf;
```

Comment:
It doesn't look hopeful that this will build on Windows. We need to guard anything platform dependent.